### DISCUSSION TITLE
How to assign different textures to SimpleMeshlayer according to corresponding coordinates?

### DISCUSSION THREAD

**Discussion by nkennyxfab** [visgl/deck.gl#7101] [Q&A]
Hi everyone, I am attempting to pass in different textures to a SimpleMeshLayer. From my understanding, you can only provide one texture per SimpleMeshLayer and this layer can render some arbitrary number of geometries according to the supplied coordinate positions. Each of those geometries share that one texture.

So, how do I assign **multiple textures** to each of those geometry, assuming that the corresponding texture will be automatically assigned to geometry's x, y coordinates? 

You can imagine that in this [SimpleMeshLayer](https://deck.gl/docs/api-reference/mesh-layers/simple-mesh-layer), each of those geometries have different textures rather than one uniform look/texture.

**Comment by ibgreen** [comment]
SimpleMeshLayer is called simple for a reason. For your use case, consider using ScenegraphLayer, it supports the glTF v2 specification.

**Comment by nkennyxfab** [comment]
Hi @ibgreen, thanks for the info. To clarify my use case even further, I am actually attempting to render very "simple" meshes, i.e. just a square polygons all across over a map area. This is where I had supplied a flat CubeGeometry() as mesh property in the SimpleMeshLayer, in the hopes of being able to provide them with corresponding textures.

If I can do that with GridLayer or GridCellLayer, that would be perfect but it doesn't seem like it can do that (if that is possible, please do let me know). As for ScenegraphLayer, I'm not quite sure if it would help me because I am not attempting to load gltf files.

**Reply by Pessimistress** [reply]
You may want to consider the IconLayer.

**Reply by nkennyxfab** [reply]
Hi @Pessimistress, I actually have considered IconLayer but I needed the icons to stay small as I zoom far away out of the map and enlarge as I zoom in, rather than staying the same size regardless of the zooming. Is there a way to control this parameter?

**Reply by Pessimistress** [reply]
[sizeUnits](https://deck.gl/docs/api-reference/layers/icon-layer#sizeunits)

**Reply by nkennyxfab** [reply]
Isn't sizeUnits just defining the units? I tried to play around with sizeMaxPixels and sizeMinPixels but doesn't seem to provide any difference.

**Reply by nkennyxfab** [reply]
As in, when I zoom out, the icon still can be seen from far away.

**Reply by Pessimistress** [reply]
`sizeUnits: 'meters'` will do exactly what you described:  

> I needed the icons to stay small as I zoom far away out of the map and enlarge as I zoom in, rather than staying the same size regardless of the zooming

**Reply by nkennyxfab** [reply]
Thanks! that works. I also have some workaround for using SimpleMeshLayer so I will evaluate which method works best. Thanks again!

**Comment by ibgreen** [comment]
> As for ScenegraphLayer, I'm not quite sure if it would help me because I am not attempting to load gltf files.

My point was that a GLTF file is a simple JSON specification that describes a scene graph AND it supports things like the UV indices that you need. 

And a simple scene graph (either using a hierarchy of luma.gl ScenegraphNodes or via a small glTF file) can be created programmatically without much effort. As straightforward as it is, it does requires some knowledge, and we don't have a lot of docs and examples for this so I agree it would probably require more effort than is reasonable for your situation.

**Reply by nkennyxfab** [reply]
That's unfortunate but thanks again for the clarification. I've applied different textures into different SimpleMeshLayers. It's a workaround for now. Do you know what's the performance penalty for doing this (if any), say like 10 layers of SimpleMeshLayers while each loading, say, 10k data points?
